File: dotnet/Platform/OpenStack.Ex/Platform_Ex.cs
using EnginX;
using OpenStack.Client;
using OpenStack.Gfx;
using System;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
#pragma warning disable CS0649, CS0169

[assembly: InternalsVisibleTo("OpenStack.GfxTests")]

namespace OpenStack;

#region Client

public class ExClientHost : Game, IClientHost {
    public ClientBase Client;
    public SceneBase Scene;
    public IPluginHost PluginHost;

    public ExClientHost(Func<ClientBase> client) {
        Client = client();
        DeviceManager = new GraphicsDeviceManager(this);
        //PluginHost = pluginHost; IPluginHost pluginHost, string title
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public T GetScene<T>() where T : SceneBase => Scene as T;

    public void SetScene(SceneBase scene) { Scene?.Dispose(); Scene = scene; Scene?.Load(); }

    protected override async Task LoadContent() {
        await base.LoadContent();
        await Client.LoadContent();
    }

    protected override async Task UnloadContent() {
        await Client.UnloadContent();
        await base.UnloadContent();
    }
}


#endregion

#region Platform

// ExObjectBuilder
// MISSING

// ExGfxSprite2D
public class ExGfxSprite2D : IOpenGfxSprite<object, object> {
    readonly ISource _source;
    readonly SpriteManager<object> _spriteManager;
    readonly ObjectSpriteManager<object, object> _objectManager;

    public ExGfxSprite2D(ISource source) {
        _source = source;
        //_spriteManager = new SpriteManager<Sprite2D>(source, new GodotSpriteBuilder());
        //_objectManager = new Object2dManager<Node, Sprite2D>(source, new GodotObjectBuilder());
    }

    public ISource Source => _source;
    public SpriteManager<object> SpriteManager => _spriteManager;
    public ObjectSpriteManager<object, object> ObjectManager => _objectManager;
    public Task<T> GetAsset<T>(object path) => _source.GetAsset<T>(path);
    public object CreateSprite(object path) => _spriteManager.CreateSprite(path).spr;
    public void PreloadSprite(object path) => throw new NotImplementedException();
    public object CreateObject(object path, object parent) => throw new NotImplementedException();
    public void PreloadObject(object path) => throw new NotImplementedException();
    public void AttachObject(AttachObjectMethod method, object source, params object[] args) => throw new NotImplementedException();
}

// ExPlatform
public class ExPlatform : Platform {
    public static readonly Platform This = new ExPlatform();
    ExPlatform() : base("EX", "EnginX") {
        GfxFactory = source => [new ExGfxSprite2D(source), null, null];
        SfxFactory = source => [new SystemSfx(source)];
    }
}

#endregion

File: Python/openstk/platforms/ex/platform_ex.py
from __future__ import annotations
import os, io, numpy as np
from openstk.core import Platform
from openstk.platforms.system import SystemSfx
from openstk.gfx import IOpenGfxSprite, IOpenGfxModel, ObjectModelBuilderBase, ObjectModelManager, MaterialBuilderBase, MaterialManager, ShaderBuilderBase, ShaderManager, TextureManager, TextureBuilderBase
from openstk.client import IClientHost
from openstk.gfx.egin import Game, GraphicsDeviceManager

# typedefs
class ISource: pass

#region Client

# ExClientHost
class ExClientHost(Game, IClientHost):
    def __init__(self, client: callable):
        super().__init__()
        # self.deviceManager = GraphicsDeviceManager(self)
        self.client: ClientBase = client()
        self.scene: SceneBase = None
        self.pluginHost: IPluginHost = None

    def getScene[T](self) -> T: return self.scene

    def setScene(self, scene: SceneBase) -> None:
        if self.scene: self.scene.dispose()
        self.scene = scene
        self.scene.load() 

    def loadContent(self) -> None:
        super().loadContent()
        self.client.loadContent()

    def unloadContent(self) -> None:
        self.client.unloadContent()
        super().unloadContent()

#endregion

#region Platform

# ExObjectBuilder
# MISSING

# ExGfxApi
class ExGfxApi(IOpenGfxApi):
    def __init__(self, source: ISource):
        self.source: ISource = source
    def attachObject(self, method: AttachObjectMethod, source: object, args: list[object]) -> object: raise NotImplementedError()

# ExGfxSprite2D
class ExGfxSprite2D(IOpenGfxSprite):
    def __init__(self, source: ISource):
        self.source: ISource = source
        # self.spriteManager: SpriteManager = SpriteManager(source, OpenGLTextureBuilder())
        # self.objectManager: ObjectSpriteManager = ObjectManager(source, OpenGLObjectModelBuilder())

    def createSprite(self, path: object, level: range = None) -> int: return self.spriteManager.createSprite(path)[0]
    def preloadSprite(self, path: object) -> None: self.textureManager.spriteManager(path)
    def createObject(self, path: object, parent: object) -> (object, dict[str, object]): raise NotImplementedError()
    def preloadObject(self, path: object) -> None: raise NotImplementedError()
    def getAsset(self, t: type, path: object) -> object: return self.source.getAsset(t, path)

# ExPlatform
class ExPlatform(Platform):
    def __init__(self):
        super().__init__('EX', 'EnginX')
        self.gfxFactory = staticmethod(lambda source: [ExGfxApi(source), ExGfxSprite2D(source), None, None, None])
        self.sfxFactory = staticmethod(lambda source: [SystemSfx(source)])
ExPlatform.This = ExPlatform()

#endregion